changes/2024-TBD_cache-across-hierarchy-keyrings/change.md

```diff
@@ -0,0 +1,613 @@
+  (this is for native language implementations only,
+  there is, currently, no CachingCMM implemented
+  in Dafny)
+- We do NOT support a shared cache between multiple
+  Searchable Encryption Configurations.
+
+We will focus on the Hierarchy Keyring unless mentioned
+otherwise. Our aim is to support a shared cache between
+multiple Hierarchy Keyrings
+as highlighted in the next sections.
+
+## Motivation
+
+The Hierarchy Keyring,
+and it's component the (Branch) Key Store,
+allow MPL Consumers to reduce their KMS Call volume
+by persisting KMS protected cryptographic materials into
+an available medium
+(currently, only a DynamoDB table is available as persistence medium).
+
+We call these cryptographic materials Branch Keys.
+
+However, an instance of the Hierarchy Keyring
+can only ever call KMS with one KMS Relationship,
+which is, at least partly,
+configured on the KMS Client determined
+at the Hierarchy Keyring's construction.
+
+By KMS Relationship, we mean any or all of the following:
+
+- KMS Configuration
+- Credentials used when creating the KMS Client, and thus
+  used when calling KMS
+- Other properties of the KMS Client, such as the region,
+  or request headers
+
+The Local Cryptographic Material Cache of
+the Hierarchy Keyring instance is then only
+populated with Branch Keys that correspond with
+that KMS relationship.
+
+Which is appropriate,
+as it is clear under what KMS relationship
+a Branch Key is accessed.
+
+However,
+the Hierarchy Keyring,
+and it's Key Store,
+have a runtime cost,
+exerting memory pressure
+and, without manual optimization,
+requiring at least 2 TLS handshakes
+when first serving a request
+(TLS to KMS & TLS to DDB).
+
+Additionally,
+the local Cryptographic Materials Cache
+exerts some runtime cost,
+particularly in a multi-threaded environment,
+when a background worker thread MAY be refreshing
+or pruning entries of the cache.
+
+For MPL Consumers that MUST work with Branch Keys
+under different KMS Relationships,
+this runtime cost adds up.
+
+These MPL Consumers MAY end up establishing
+a LRU Cache of Hierarchy Keyrings.
+Which, while workable, is sub-optimal,
+and clearly makes the Hierarchy Keyring,
+in these conditions,
+"Hard to Use".
+
+The objective, with these changes,
+is to make the Hierarchy Keyring
+"Easy to Use" in a multiple KMS Relationship
+environment.
+
+## Summary
+
+<!-- TODO: Remove SIM before merging to mainline -->
+
+Relevant [sim](https://sim.amazon.com/issues/CrypTool-5311).
+
+To improve the MPL Consumer data key caching experience,
+when using the Hierarchy Keyring,
+we need to allow caching across Key Stores/KMS Clients/KMS Keys.
+
+To facilitate Caching across Key Stores/KMS Clients/KMS Keys,
+we MUST break the Cryptographic Materials Cache (CMC)
+out of the Hierarchy Keyring.
+
+By allowing MPL Consumers to provide an already initialized CMC
+to the Hierarchy Keyring at construction,
+the CMC MAY cache Branch Keys protected by different
+KMS Relationships.
```

Comment:
Adding this cache MUST be an optional parameter, right? Otherwise current customers of the keyring will get compile errors saying, "must provide cache in initialization" or something like that. 
```suggestion
By allowing MPL Consumers to optionally provide an initialized CMC
to the Hierarchy Keyring at construction,
the CMC MAY cache Branch Keys protected by different
KMS Relationships.
```

Reply:
Yes, you're right. It is an optional parameter that the customer can use.